### PR TITLE
etag can be passed in the if-none-match in the header on getobject api

### LINKED ISSUE #5967: Add ETag getObject with If-None-Match
## What

Add an optional [`If-None-Match: "ETag"`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match#syntax) header to `getObject` requests.  When an ETag is passed, if that ETag is current for the object on that path then return 304 Not Modified.

(Note that for a non-branch ref, we expect the ETag always to match!)

## Use-case

My server reads configuration from an object stored on lakeFS.  Periodically it needs to refresh the configuration.  So it calls getObject.  Ideally it would be able to avoid reading the data during a refresh.

Even better: if lakeFS versions _before_ we add this feature simply ignore the header, my server will work with them too!

### PR DESCRIPTION
**Pull request by vishalsanfran**
Closes #5967

## Change Description

Add an optional [If-None-Match: "ETag"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match#syntax) header to getObject requests. When an ETag is passed, if that ETag is current for the object on that path then return 304 Not Modified.

If the ETag is not current, return the object as it was being originally done. This ensures that this is not a breaking change.

### Testing Details

Added a unit test that checks if a 304 is returned (pkg/api/controller_test.go)

### Breaking Change?

If-None-Match is an optional parameter and existing clients will need to be updated to pass this parameter. So, it should not be a breaking change.

### CODE REVIEW

**Comment by arielshaqed** [comment]
Hi @vishalsanfran ,
Thanks for your contribution to lakeFS! And sorry we missed this one. Other than a leftover debugging print it looks neat.

I am on favour of this change but we will have to consider it carefully.

It affects core semantics, so we would want to be sure that we can continue to support it - especially as we near 1.0.

I'd like to explain why we don't have this functionality. Obviously you can already get it, either by running statObject before getObject or through the s3 gateway API. At issue with providing it directly on GET are two things.

First, this is "only" a caching API. Unfortunately conditional GET/POST APIs have a record of being used for synchronization - which is where they fail. I believe we might reduce this worry by documenting this option better in the openapi spec.

Second, I am personally in favour of removing this pair of APIs entirely in 1.0, in favour of presigned URLs and/or direct access to storage. In this scenario getObject would no longer exist, you'd statObject and then read from a URL it returns.

Could you share use cases for this? I'm particularly interested in anything that we cannot easily support without having this feature. Thanks!

File: pkg/api/controller_test.go [hunk 1]
```diff
@@ -2399,6 +2399,20 @@ func TestController_ObjectsGetObjectHandler(t *testing.T) {
 			t.Errorf("expected to get \"%s\" storage class, got %#v", expensiveString, properties)
 		}
 	})
+
+	t.Run("get object not modified", func(t *testing.T) {
+		etag_input := "\"" + blob.Checksum + "\""
```
Comment (nopcoder) [inline pkg/api/controller_test.go:2404]: Go naming etag_input => etagInput

**Review by nopcoder — COMMENTED** [review]
In addition to @arielshaqed feedback - small comments about the code

**Comment by vishalsanfran** [comment]
@arielshaqed thank you for the comment.
The usecase was to perform an optimized retrieval of an object just as it was described in the issue #5967 
It should still be a non breaking change since this is an optional param.

**Comment by arielshaqed** [comment]
> @arielshaqed thank you for the comment. The usecase was to perform an optimized retrieval of an object just as it was described in the issue #5967 It should still be a non breaking change since this is an optional param.

Yeah, I know the guy who submitted that issue 😕 . This started the chain of thoughts: we either eliminate the call or we need to enhance it. But there are so many things missing from it!
* Conditional upload
* Incompatibility with OpenAPI, unless the content is uploaded as a file named "content"
* (No) support for multipart upload; maximal size you can upload depends on the backing store
* Limited performance: lakeFS becomes a data proxy, so it uses 2x the bandwidth that the client uses
* One of only 2 data operations in a service that handles only metadata (the other is getObject). This is three architectural cause of the other issues.
* Limited support for storage class. This was actually the first feature I added to lakeFS.

This last one is instructive. I added it for a single user who wanted it. Because at the time we had no direct upload, they had no recourse but to wait for our feature. So for any feature available on uploads to the backing store, we need to add that same feature to lakeFS. But because we support multiple backing stores, we need to work harder than AWS, Microsoft and Google combined, and we will never be done.

A better plan may be to provide only the 3-step upload prices, and then write simple client libraries to upload.

So there is a very real chance that we will get rid of putObject and getObject for 1.0. To be very clear! **I estimate it will probably stay, but the probability of it going away is say 0.4**.

Could we delay this one for 2 weeks, until we understand the future of these calls?

**Comment by arielshaqed** [comment]
This is not a 1.0.0 blocker.  Additionally we decided to _keep_ the getObject API, and it is fairly web-by to have such a facility there.  **Unblocking** this, and sorry for the delays!

File: pkg/api/controller.go [hunk 2]
```diff
@@ -3611,6 +3611,16 @@ func (c *Controller) GetObject(w http.ResponseWriter, r *http.Request, repositor
 		return
 	}
 
+	// check ETag if not modified
+	etag := httputil.ETag(entry.Checksum)
+	fmt.Println(r.Header.Get("If-None-Match"))
+	fmt.Println(etag)
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/api/controller.go:3617]: @vishalsanfran gentle "ping" to remove these :-)

**Review by arielshaqed — COMMENTED** [review]
Thanks @vishalsanfran and sorry for delaying this PR!  We are good to go, please see @nopcoder 's comments :-)

**Comment by arielshaqed** [comment]
Also @vishalsanfran could you please rebase this PR onto our `main`?

**Comment by nopcoder** [comment]
Hi @vishalsanfran,

Thank you for your contribution and for taking the time to put together this pull request.

Upon review, we've noted that the implementation makes use of the `if-none-match` parameter. As part of the changes introduced in the v1 stable API, this parameter has been marked as deprecated, and we plan to phase out its support in the near future.

Due to this, we will close this pull request at this time. We appreciate the effort and quality of the work you've done here, and we would love to see more contributions from you in the future.

**Comment by nopcoder** [comment]
I've reopened the pull request as well as the associated issue. We've decided to retract the deprecation notice for the if-none-match argument. Initially, we believed that our future needs could be met through alternative capabilities, but upon further consideration, we've recognized its necessity.

**Pull request closed without merging**